### PR TITLE
use brew bundle and brew services instead of custom brew functions

### PR DESCRIPTION
**Pull request by AlJohri**
closes https://github.com/thoughtbot/laptop/issues/401

### CODE REVIEW

**Comment by mikemcquaid** [comment]
:+1:

**Comment by gohanlon** [comment]
I contributed to several of these Bash functions, and have no love for them. If I did want to defend them I'd mention that they do work and have been reasonably stable for a while now, and using them avoids an external dependency. But, Brew Bundle having official Homebrew support makes it more worth considering, IMO. (It does have "official" Homebrew support, right?)

One important note to discuss, though, about the addition of `brew services`: Laptop used to use `brew services`, but had at least [one issue related to it](https://github.com/thoughtbot/laptop/issues/243) and in the process of resolving that issue, we noticed that [`brew services` had been deprecated](https://github.com/Homebrew/homebrew/commit/c0b99c031fec0a2096b85a458d7b42e3627560d8) and [moved away from it](https://github.com/thoughtbot/laptop/pull/254).

@mikemcquaid What's the status of `brew services`? Is it advisable to use?

**Comment by mikemcquaid** [comment]
> (It does have "official" Homebrew support, right?)

Yes.

> What's the status of brew services? Is it advisable to use?

It's been moved back into the Homebrew organisation (so is officially supported/recommended) but is not  included by default in Homebrew itself. It'll shortly be supported by `homebrew-bundle` too: https://github.com/Homebrew/homebrew-bundle/pull/96

**Comment by blackjid** [comment]
Is this working for you @AlJohri???  I like it..

File: mac [hunk 1]
```diff
@@ -131,31 +82,40 @@ fi
...
+EOF
+)
+
+fancy_echo "Installing Homebrew Packages ..."
+  brew tap 'homebrew/bundle'
+  echo "$BREWFILE" | brew bundle -v --file=/dev/stdin
+  brew unlink openssl && brew link openssl --force
+  brew services restart jack
+  brew services restart redis
+  brew services restart postgresql
```
Comment (mikemcquaid) [inline mac]: You can integrate this into the Brewfile now.

File: mac [hunk 2]
```diff
@@ -131,31 +82,40 @@ fi
...
+brew 'libyaml'
+brew 'heroku-toolbelt'
+brew 'rcm'
+EOF
+)
+
+fancy_echo "Installing Homebrew Packages ..."
+  brew tap 'homebrew/bundle'
+  echo "$BREWFILE" | brew bundle -v --file=/dev/stdin
+  brew unlink openssl && brew link openssl --force
```
Comment (mikemcquaid) [inline mac]: This is a bad idea. On 10.11 the OpenSSL headers are removed but library remains. Given Clang's default paths this means things will use the Homebrew headers but link against the system libraries. At best, it's insecure. At worst, it'll cause random, hard-to-debug runtime failures.

Comment (AlJohri) [inline mac]: Thanks for the comment! I agree and don't use this on my personal system. I meant for this PR to be exactly the current functionality.

Comment (mikemcquaid) [inline mac]: :+1:

File: mac (hunk 1 above)
Comment (mikemcquaid) [inline mac]: e.g. `brew 'postgresql', restart_service: true`

**Comment by croaky** [comment]
@AlJohri Sorry for the delay on this. Would you be up for rebasing?

It looks good to me, with two small cleanups:

* Could you sort the Brewfile `brew` commands alphabetically?
* It looks like there's a `brew services restart jack` in there that could be removed.

File: mac [hunk 3]
```diff
@@ -131,31 +82,40 @@ fi
...
-brew_install_or_upgrade 'openssl'
-brew unlink openssl && brew link openssl --force
-brew_install_or_upgrade 'libyaml'
+BREWFILE=$(cat <<'EOF'
+tap 'thoughtbot/formulae'
+tap 'homebrew/services'
+brew 'zsh'
+brew 'git'
+brew 'postgres'
+brew 'redis'
```
Comment (mikemcquaid) [inline mac]: You can do `brew "redis", restart_service: true` now.

File: mac [hunk 4]
```diff
@@ -131,31 +82,40 @@ fi
...
 
-brew_install_or_upgrade 'openssl'
-brew unlink openssl && brew link openssl --force
-brew_install_or_upgrade 'libyaml'
+BREWFILE=$(cat <<'EOF'
+tap 'thoughtbot/formulae'
+tap 'homebrew/services'
+brew 'zsh'
+brew 'git'
+brew 'postgres'
```
Comment (mikemcquaid) [inline mac]: You can do `brew "redis", restart_service: true` now.

File: mac [hunk 5]
```diff
@@ -131,31 +82,40 @@ fi
...
+brew 'rbenv'
+brew 'ruby-build'
+brew 'libyaml'
+brew 'heroku-toolbelt'
+brew 'rcm'
+EOF
+)
+
+fancy_echo "Installing Homebrew Packages ..."
+  brew tap 'homebrew/bundle'
```
Comment (mikemcquaid) [inline mac]: You don't need to do this any more; it's done automatically.

File: mac (hunk 4 above)
Comment (mikemcquaid) [inline mac]: Also, I recommend doing `postgresql` which is the full, non-aliased name.

**Comment by mikemcquaid** [comment]
> Could you sort the Brewfile brew commands alphabetically?

You'll need to leave the `tap`s at the top, though.

**Comment by croaky** [comment]
Right, after the taps.

**Comment by AlJohri** [comment]
Sure thing.

**Comment by AlJohri** [comment]
wouldn't you want to keep them in the same order with comments replacing the fancy_echos?

for example:

```
BREWFILE=$(cat <<'EOF'
tap 'thoughtbot/formulae'
tap 'homebrew/services'

# Unix Tools

brew 'ctags'
brew 'git'
brew 'openssl'
brew 'rcm'
brew 'reattach-to-user-namespace'
brew 'the_silver_searcher'
brew 'tmux'
brew 'vim'
brew 'zsh'

# Heroku

brew 'heroku-toolbelt'
brew 'parity'
```

also how do I deal with ```brew_install_or_upgrade 'libyaml' # should come after openssl```?

**Comment by mikemcquaid** [comment]
> also how do I deal with brew_install_or_upgrade 'libyaml' # should come after openssl?

@AlJohri That shouldn't matter any more.

**Comment by croaky** [comment]
@AlJohri Sure, that comment style looks good to me.

`# should come after openssl` comment can be deleted, but still think we want to `brew 'libyaml'` for Ruby.

**Comment by mikemcquaid** [comment]
> brew 'libyaml' for Ruby.

Yup (and openssl, obviously)

**Comment by mikemcquaid** [comment]
@croaky Irrelevant to this PR but you may be interested in my/GitHub's solution to the same thoughtbot/laptop problem: https://github.com/mikemcquaid/strap/blob/master/bin/strap.sh. It's under a liberal enough license that you should feel free to copy/paste bits anything you want :wink:

**Comment by croaky** [comment]
@mikemcquaid Awesome, thanks! Will take a look.

**Comment by croaky** [comment]
Looking to get this in so I opened https://github.com/thoughtbot/laptop/pull/451 but if you're able to rebase and make the same changes here, I'd rather you had the credit in the commit history!

**Comment by croaky** [comment]
Closing in favor of https://github.com/thoughtbot/laptop/pull/451 in the interest of getting this into master.

**Pull request closed without merging**